inference-engine/src/readers/ir_reader/ie_ir_parser.cpp

```diff
@@ -196,7 +196,135 @@ std::shared_ptr<ICNNNetwork> V10Parser::parse(const pugi::xml_node& root, std::i
+        noOfChannels = inputDims[1];
+        width = inputDims[1];
+        height = inputDims[0];
+    } else if (inputDims.size() == 3) {
+        width = inputDims[2];
+        height = inputDims[1];
+        noOfChannels = inputDims[0];
+    } else if (inputDims.size() == 4) {
+        width = inputDims[3];
+        height = inputDims[2];
+        noOfChannels = inputDims[1];
+    } else if (inputDims.size() == 5) {
+        width = inputDims[4];
+        height = inputDims[3];
+        noOfChannels = inputDims[2];
+    }
+
+    PreProcessInfo& pp = preProcessInput->getPreProcess();
+    pp.init(noOfChannels);
+
+    auto meanSegmentPrecision = GetPrecisionAttr(ppNode, "mean-precision", Precision::UNSPECIFIED);
+    if (!meanSegmentPrecision || meanSegmentPrecision == Precision::MIXED)
+        THROW_IE_EXCEPTION << "mean blob defined without specifying precision.";
+
+    ResponseDesc resp;
+    InferenceEngine::PreProcessChannel::Ptr preProcessChannel;
+
+    int lastChanNo = -1;
+    std::unordered_set<int> idsForMeanImage;
+
+    FOREACH_CHILD(chan, ppNode, "channel") {
+        int chanNo = GetIntAttr(chan, "id", lastChanNo + 1);
+        if (chanNo >= static_cast<int>(noOfChannels) || chanNo < 0) {
+            THROW_IE_EXCEPTION << "Pre-process channel id invalid: " << chanNo;
+        }
+        lastChanNo = chanNo;
+        preProcessChannel = pp[chanNo];
+
+        auto meanNode = chan.child("mean");
+        if (!meanNode.empty()) {
+            if (!meanNode.attribute("size")) {
+                THROW_IE_EXCEPTION << "mean should have the attribute: size";
+            }
+            if (meanNode.attribute("size")) {
+                idsForMeanImage.insert(chanNo);
+                size_t size = static_cast<size_t>(GetIntAttr(meanNode, "size"));
+                size_t offset = static_cast<size_t>(GetIntAttr(meanNode, "offset"));
+                if (width * height * meanSegmentPrecision.size() != size) {
+                    THROW_IE_EXCEPTION << "mean blob size mismatch expected input, got: " << size
+                                       << " extpecting " << width << " x " << height << " x "
+                                       << meanSegmentPrecision.size();
+                }
+                preProcessChannel->meanData = make_blob_with_precision(TensorDesc(meanSegmentPrecision, {height, width}, Layout::HW));
+                preProcessChannel->meanData->allocate();
+                auto lockedMem = preProcessChannel->meanData->buffer();
+                char* data = lockedMem.as<char *>();
+                binStream.seekg(offset, std::ios::beg);
+                binStream.read(data, size);
+            }
+        }
+    }
+
+    if (idsForMeanImage.size() == noOfChannels) {
+        pp.setVariant(MEAN_IMAGE);
+    } else if (idsForMeanImage.size() == 0) {
+        pp.setVariant(NONE);
+    } else {
+        std::string validMeanImageIds = "";
+        for (auto id : idsForMeanImage) {
```

Comment:
Why do we need cycle if we throw an exception later?

Reply:
We generate the right message for an exception in this cycle.